config/config.go

```diff
@@ -136,37 +136,75 @@ const (
 )
 
 type Metrics struct {
-	Type   MetricsType `mapstructure:"type"`
-	Influx Influx      `mapstructure:"influx"`
+	Influx     InfluxMetrics     `mapstructure:"influx"`
+	Prometheus PrometheusMetrics `mapstructure:"prometheus"`
 }
 
 func (cfg *Metrics) validateAndLog() {
-	log.Infof("config.metrics.type: %s", cfg.Type)
-	switch cfg.Type {
-	case MetricsNone:
-	case MetricsInflux:
-		cfg.Influx.validateAndLog()
-	default:
-		log.Fatalf(`invalid config.metrics.type: %s. It must be "none" or "influx"`, cfg.Type)
+	var metricsEnabled bool = false
+	if cfg.Influx.Enabled {
+		metricsEnabled = cfg.Influx.validateAndLogMetricsData()
+	}
+	if cfg.Prometheus.Enabled {
+		metricsEnabled = cfg.Prometheus.validateAndLogMetricsData() || metricsEnabled
+	}
+	if !metricsEnabled {
+		log.Fatalf(`No metrics correctly specified in configuration file`)
 	}
 }
 
-type MetricsType string
-
-const (
-	MetricsNone   MetricsType = "none"
-	MetricsInflux MetricsType = "influx"
-)
+type MetricsConfig interface {
```

Comment:
Do you use this interface definition anywhere?

Reply:
The `MetricsConfig` is only used in this file to make `cfg.Influx` and `cfg.Prometheus` implement this interface so they have a `validateAndLogMetricsData()` function to validate we had enough info in the configuration files (or environment variables) to init the prometheus or influx metrics
```
143 func (cfg *Metrics) validateAndLog() {
144     var metricsEnabled bool = false
145     if cfg.Influx.Enabled {
146         metricsEnabled = cfg.Influx.validateAndLogMetricsData()
147     }
148     if cfg.Prometheus.Enabled {
149         metricsEnabled = cfg.Prometheus.validateAndLogMetricsData() || metricsEnabled
150     }
151     if !metricsEnabled {
152         log.Fatalf(`No metrics correctly specified in configuration file`)
153     }
154 }
155
156 type MetricsConfig interface {
157     validateAndLogMetricsData() bool
158 }
159
160 +--  7 lines: type InfluxMetrics struct {-------------------------------------------------------------------------------
167
168 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
169     if influxMetricsConfig.Host != "" {
170         influxMetricsConfig.Enabled = false
171         log.Fatalf(`Despite being enabled, influx metrics came with no host info: config.metrics.influx.host = "".`)
172     }
173     if influxMetricsConfig.Database != "" {
174         influxMetricsConfig.Enabled = false
175         log.Fatalf(`Despite being enabled, influx metrics came with no database info: config.metrics.influx.database = "".`)
176     }
177     if influxMetricsConfig.Enabled {
178         log.Infof("config.metrics.influx.host: %s", influxMetricsConfig.Host)
179         log.Infof("config.metrics.influx.database: %s", influxMetricsConfig.Database)
180     }
181     return influxMetricsConfig.Enabled
182 }
183
184 +--  7 lines: type PrometheusMetrics struct {---------------------------------------------------------------------------
191
192 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
193     if promMetricsConfig.Port == 0 {
194         promMetricsConfig.Enabled = false
195         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty port number: config.metrics.prometheus.port = 0`)
196     }
197     if promMetricsConfig.Namespace != "" {
198         promMetricsConfig.Enabled = false
199         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty name space: config.metrics.prometheus.namespace = \"\".`)
200     }
201     if promMetricsConfig.Subsystem != "" {
202         promMetricsConfig.Enabled = false
203         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty subsystem value: config.metrics.prometheus.subsystem = \"\".`)
204     }
205     if promMetricsConfig.Enabled {
206         log.Infof("config.metrics.prometheus.namespace: %s", promMetricsConfig.Namespace)
207         log.Infof("config.metrics.prometheus.subsystem: %s", promMetricsConfig.Subsystem)
208     }
209     return promMetricsConfig.Enabled
210 }
config/config.go
```

Reply:
I did a search on the PR and I am not seeing it referenced anywhere. Can you point me to where you are using it?

Reply:
Sure. Objects `InfluxMetrics` and `PrometheusMetrics` are meant to implement this interface. Do you think an `interface` is too much? Want me to go with methods instead? The objective of it being an interface was to keep consistensy with more than one metrics engine.
```
168 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool
  .
  .
  .
192 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool
config/config.go
```

Get used:
```
143 func (cfg *Metrics) validateAndLog() {
144     var metricsEnabled bool = false
145     if cfg.Influx.Enabled {
146  >>     metricsEnabled = cfg.Influx.validateAndLogMetricsData()
147     }
148     if cfg.Prometheus.Enabled {
149  >>     metricsEnabled = cfg.Prometheus.validateAndLogMetricsData() || metricsEnabled
150     }
151     if !metricsEnabled {
152         log.Fatalf(`No metrics correctly specified in configuration file`)
153     }
154 }
config/config.go
```

Reply:
What I mean do you every have a variable that is of type `MetricsConfig`? The two engines do implement the method, but that is true if you define the interface or not. Having a definition for `MetricsConfig` does not guarantee that any particular object implements it.

Reply:
I meant `Influx InfluxMetrics` and `Prometheus PrometheusMetrics` to implement the `MetricsConfig` interface when defining their implementations of `validateAndLogMetricsData()`. Do you want me to get rid of the interface and just leave those  `InfluxMetrics.validateAndLogMetricsData()` and `PrometheusMetrics.validateAndLogMetricsData()` functions as methods?
```
138 type Metrics struct {
139     Influx     InfluxMetrics     `mapstructure:"influx"`
140     Prometheus PrometheusMetrics `mapstructure:"prometheus"`
141 }
142
143 +--  8 lines: func (cfg *Metrics) validateAndLog() {------------------------------
151
152 type MetricsConfig interface {
153     validateAndLogMetricsData() bool
154 }
155
156 type InfluxMetrics struct {
157     Host     string `mapstructure:"host"`
158 +--  3 lines: Database string `mapstructure:"database"`----------------------------
161     Enabled  bool   `mapstructure:"enabled"`
162 }
163
164 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
165     if influxMetricsConfig.Host != "" {
166 +-- 11 lines: influxMetricsConfig.Enabled = false----------------------------------
177     return influxMetricsConfig.Enabled
178 }
179
180 type PrometheusMetrics struct {
181     Port             int    `mapstructure:"port"`
182 +--  3 lines: Namespace        string `mapstructure:"namespace"`-------------------
185     Enabled          bool   `mapstructure:"enabled"`
186 }
187
188 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
189     if promMetricsConfig.Port == 0 {
190 +-- 15 lines: promMetricsConfig.Enabled = false------------------------------------
205     return promMetricsConfig.Enabled
206 }
config/config.go
```